File: python/tests/bulk_vectors.py
"""Shared batch behavior and exact value-free metric bytes."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

import sde


class Rollback(Exception):
    pass


def drive_bulk_vector(case: Path, model: Any, placement: Any, engines: Any) -> None:
    want = json.loads((case / "bulk.json").read_text())
    recorder = sde.Recorder(model.version)
    session = sde.Session(model, placement, engines, recorder=recorder)
    errors: list[str] = []

    def run(steps: list[dict[str, Any]]) -> None:
        for step in steps:
            try:
                if step["op"] == "transaction":
                    try:
                        with session.transaction("Reading"):
                            run(step["body"])
                            if step.get("rollback"):
                                raise Rollback
                    except Rollback:
                        pass
                else:
                    rows = step["rows"] * step.get("repeat", 1)
                    session.save_many(step.get("entity", "Reading"), rows)
            except sde.SdeError as exc:
                assert type(exc).__name__ == step.get("error")
                errors.append(type(exc).__name__)
            else:
                assert "error" not in step

    run(want["operations"])
    engine = next(iter(engines.values()))
    assert engine.recorded.calls == want["calls"]
    assert {name: db.tables for name, db in engines.items()} == want["tables"]
    assert errors == want["errors"]
    window = recorder.roll()
    metrics = {
        "shapes": []
        if window is None
        else [
            {
                key: getattr(stats, key)
                for key in ("shape_id", "entity", "group", "kind", "calls", "rows", "errors")
            }
            for stats in window.shapes
        ],
        "copies": []
        if window is None
        else [
            {key: getattr(stats, key) for key in ("group", "materialization", "writes", "failures")}
            for stats in window.fanned
        ],
    }
    assert metrics == want["metrics"]
    assert sde.canonical_bytes(metrics).hex() == want["metrics_hex"]

File: typescript/tests/_bulk-vectors.ts
/** Shared batch call order, data and exact value-free metric bytes. */
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { expect } from 'vitest'
import { canonicalBytes, Recorder, Session, SdeError, type LogicalModel, type PlacementMap, type Row } from '../src/index.js'
import type { MemoryEngine } from '../src/testing/memory.js'
interface Step {
  op: string; rows?: Row[]; repeat?: number; entity?: string; body?: Step[]; rollback?: boolean; error?: string
}
class Rollback extends Error {}
export async function driveBulkVector(dir: string, model: LogicalModel, map: PlacementMap, engines: Record<string, MemoryEngine>): Promise<void> {
  const want = JSON.parse(readFileSync(join(dir, 'bulk.json'), 'utf8')) as {
    operations: Step[]; calls: unknown[]; tables: unknown; errors: string[]; metrics: unknown; metrics_hex: string
  }
  const recorder = new Recorder(model.version)
  const session = await Session.open(model, map, engines, { recorder })
  const errors: string[] = []
  async function run(steps: Step[]) {
    for (const step of steps) {
      try {
        if (step.op === 'transaction') {
          try {
            await session.transaction(['Reading'], async () => {
              await run(step.body!); if (step.rollback) throw new Rollback()
            })
          } catch (error) { if (!(error instanceof Rollback)) throw error }
        } else {
          const rows = Array.from({ length: step.repeat ?? 1 }, () => step.rows!).flat()
          await session.saveMany(step.entity ?? 'Reading', rows)
        }
      } catch (error) {
        expect(error).toBeInstanceOf(SdeError)
        expect((error as Error).name).toBe(step.error)
        errors.push((error as Error).name)
        continue
      }
      expect(step.error).toBeUndefined()
    }
  }
  await run(want.operations)
  expect(Object.values(engines)[0]!.recorded.calls).toEqual(want.calls)
  expect(Object.fromEntries(Object.entries(engines).map(([name, db]) => [name, db.tables]))).toEqual(want.tables)
  expect(errors).toEqual(want.errors)
  const window = recorder.roll()
  const metrics = {
    shapes: window?.shapes.map(stats => ({ shape_id: stats.shapeId, entity: stats.entity, group: stats.group,
      kind: stats.kind, calls: stats.calls, rows: stats.rows, errors: stats.errors })) ?? [],
    copies: window?.fanned.map(stats => ({ group: stats.group, materialization: stats.materialization,
      writes: stats.writes, failures: stats.failures })) ?? [],
  }
  expect(metrics).toEqual(want.metrics)
  expect(Buffer.from(canonicalBytes(metrics)).toString('hex')).toBe(want.metrics_hex)
}
